File: tool/src/main/java/com/lightphone/imessage/domain/codec/IMessageCodec.java
package com.lightphone.imessage.domain.codec;

import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.cert.X509Certificate;
import com.lightphone.imessage.domain.codec.MessagePayload;

/**
 * Interface for encoding and decoding iMessage envelopes.
 * Handles encryption/decryption and signing/verification of message payloads.
 */
public interface IMessageCodec {
    /**
     * Encodes a message payload into an encrypted, signed envelope.
     *
     * @param payload Message payload to encode
     * @param recipientKey RSA-2048 public key of recipient
     * @param senderKey ECDSA P-256 private key of sender
     * @return Result containing binary envelope or failure
     */
    Result<byte[]> encodeEnvelope(
        MessagePayload payload,
        PublicKey recipientKey,
        PrivateKey senderKey
    );

    /**
     * Decodes and verifies an encrypted, signed envelope into a message payload.
     *
     * @param envelope Binary envelope
     * @param recipientKey RSA-2048 private key of recipient
     * @param senderCert X509 certificate of sender
     * @return Result containing MessagePayload or failure
     */
    Result<MessagePayload> decodeEnvelope(
        byte[] envelope,
        PrivateKey recipientKey,
        X509Certificate senderCert
    );

    /**
     * Encodes a Plist value to binary Plist bytes.
     *
     * @param value PlistValue to encode
     * @return Result containing binary Plist bytes or failure
     */
    Result<byte[]> encodePlist(PlistValue value);

    /**
     * Decodes binary Plist bytes into a PlistValue.
     *
     * @param bytes Binary Plist data
     * @return Result containing PlistValue or failure
     */
    Result<PlistValue> decodePlist(byte[] bytes);
}

File: tool/src/main/java/com/lightphone/imessage/domain/codec/IMessageCodec.kt
package com.lightphone.imessage.domain.codec

import java.security.PrivateKey
import java.security.PublicKey
import java.security.cert.X509Certificate

/**
 * Interface for encoding and decoding iMessage envelopes.
 *
 * Handles authenticated encryption (AES-256-GCM), key wrapping (RSA-2048-OAEP-SHA256), and message
 * signing/verification (ECDSA P-256 + SHA-256) of message payloads.
 *
 * All methods return [kotlin.Result] so that callers can distinguish crypto/parse failures from
 * success without exception plumbing.
 */
interface IMessageCodec {
    /**
     * Encode [payload] into an encrypted, signed envelope.
     *
     * @param payload Message payload to encode.
     * @param recipientKey RSA-2048 public key of the recipient (used to wrap the AES key).
     * @param senderKey ECDSA P-256 private key of the sender (used to sign the envelope).
     * @return [Result] containing the binary envelope bytes, or a failure with the underlying
     * cause.
     */
    fun encodeEnvelope(
            payload: MessagePayload,
            recipientKey: PublicKey,
            senderKey: PrivateKey,
    ): Result<ByteArray>

    /**
     * Decode and verify [envelope] into a [MessagePayload].
     *
     * The implementation MUST verify the ECDSA signature BEFORE any AES unwrap/decrypt happens, and
     * MUST reject unsupported envelope versions.
     *
     * @param envelope Binary envelope bytes.
     * @param senderCert X.509 certificate of the sender (public key used to verify signature).
     * @param recipientKey RSA-2048 private key of the recipient (used to unwrap the AES key).
     * @return [Result] containing the decoded [MessagePayload], or a failure with the underlying
     * cause.
     */
    fun decodeEnvelope(
            envelope: ByteArray,
            senderCert: X509Certificate,
            recipientKey: PrivateKey,
    ): Result<MessagePayload>
}
